x/auth/middleware/branch_store_test.go

```diff
@@ -0,0 +1,113 @@
+package middleware_test
+
+import (
+	"context"
+	"fmt"
+	"math"
+
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	"github.com/cosmos/cosmos-sdk/simapp"
+	"github.com/cosmos/cosmos-sdk/testutil/testdata"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/types/tx"
+	"github.com/cosmos/cosmos-sdk/x/auth/middleware"
+	minttypes "github.com/cosmos/cosmos-sdk/x/mint/types"
+)
+
+var blockMaxGas = uint64(simapp.DefaultConsensusParams.Block.MaxGas)
+
+func (s *MWTestSuite) TestBranchStore() {
+	testcases := []struct {
+		name         string
+		gasToConsume uint64 // gas to consume in the msg execution
+		panicTx      bool   // panic explicitly in tx execution
+		expErr       bool
+	}{
+		{"less than block gas meter", 10, false, false},
+		{"more than block gas meter", blockMaxGas, false, true},
+		{"more than block gas meter", uint64(float64(blockMaxGas) * 1.2), false, true},
+		{"consume block gas when paniced", 10, true, true},
+	}
+
+	for _, tc := range testcases {
+		s.Run(tc.name, func() {
+			ctx := s.SetupTest(true).WithBlockGasMeter(sdk.NewGasMeter(blockMaxGas)) // setup
+			txBuilder := s.clientCtx.TxConfig.NewTxBuilder()
+
+			// tx fee
+			feeCoin := sdk.NewCoin("atom", sdk.NewInt(150))
+			feeAmount := sdk.NewCoins(feeCoin)
+
+			// test account and fund
+			priv1, _, addr1 := testdata.KeyTestPubAddr()
+			err := s.app.BankKeeper.MintCoins(ctx, minttypes.ModuleName, feeAmount)
+			s.Require().NoError(err)
+			err = s.app.BankKeeper.SendCoinsFromModuleToAccount(ctx, minttypes.ModuleName, addr1, feeAmount)
+			s.Require().NoError(err)
+			s.Require().Equal(feeCoin.Amount, s.app.BankKeeper.GetBalance(ctx, addr1, feeCoin.Denom).Amount)
+
+			// testMsgTxHandler is a test txHandler that handles one single TestMsg,
+			// consumes the given `tc.gasToConsume`, and sets the bank store "ok" key to "ok".
+			var testMsgTxHandler = customTxHandler{func(ctx context.Context, req tx.Request) (tx.Response, error) {
+				msg, ok := req.Tx.GetMsgs()[0].(*testdata.TestMsg)
+				if !ok {
+					return tx.Response{}, fmt.Errorf("expected %T, got %T", (*testdata.TestMsg)(nil), msg)
+				}
+
+				sdkCtx := sdk.UnwrapSDKContext(ctx)
+				sdkCtx.KVStore(s.app.GetKey("bank")).Set([]byte("ok"), []byte("ok"))
+				sdkCtx.GasMeter().ConsumeGas(tc.gasToConsume, "TestMsg")
+				if tc.panicTx {
+					panic("panic in tx execution")
+				}
+				return tx.Response{}, nil
+			}}
+
+			txHandler := middleware.ComposeMiddlewares(
+				testMsgTxHandler,
+				middleware.NewTxDecoderMiddleware(s.clientCtx.TxConfig.TxDecoder()),
+				middleware.GasTxMiddleware,
+				middleware.RecoveryTxMiddleware,
+				middleware.DeductFeeMiddleware(s.app.AccountKeeper, s.app.BankKeeper, s.app.FeeGrantKeeper),
```

Comment:
Can we also add the IncrementSeqMiddleware and check that sequence is always bumped?

Reply:
done